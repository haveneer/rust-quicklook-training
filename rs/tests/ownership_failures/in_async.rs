// edition:2018

use futures::executor::block_on;
use std::time::Duration;

async fn async_main() {
    let mut numbers = vec![1; 100];
    let sum_future = sum(&numbers);
    std::thread::sleep(Duration::from_millis(100));
    add(&mut numbers); //~ cannot borrow `numbers` as mutable
    println!("The sum is {}", sum_future.await);
}

fn add(numbers: &mut Vec<i32>) {
    numbers.push(42);
}

async fn sum(numbers: &Vec<i32>) -> i32 {
    let iter = numbers.iter();
    std::thread::sleep(Duration::from_millis(200));
    iter.sum()
}

fn main() {
    block_on(async_main());
}
