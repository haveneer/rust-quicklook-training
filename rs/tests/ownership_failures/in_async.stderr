error[E0502]: cannot borrow `numbers` as mutable because it is also borrowed as immutable
  --> tests/ownership_failures/in_async.rs:10:9
   |
8  |     let sum_future = sum(&numbers);
   |                          -------- immutable borrow occurs here
9  |     std::thread::sleep(Duration::from_millis(100));
10 |     add(&mut numbers); //~ cannot borrow `numbers` as mutable
   |         ^^^^^^^^^^^^ mutable borrow occurs here
11 |     println!("The sum is {}", sum_future.await);
   |                               ---------- immutable borrow later used here
